smart_open/smart_open_lib.py

```diff
@@ -383,10 +383,14 @@ def _open_binary_stream(uri, mode, **kw):
         return smart_open_s3.open(uri.bucket.name, uri.name, mode, **kw), uri.name
     elif hasattr(uri, 'read'):
         # simply pass-through if already a file-like
-        filename = '/tmp/unknown'
+        # we need to return something as the file name, but we don't know what
+        # so we probe for uri.name (e.g., this works with open() or tempfile.NamedTemporaryFile)
+        # if the value ends with COMPRESSED_EXT, we will note it in _compression_wrapper()
+        # if there is no such an attribute, we return "unknown" - this effectively disables any compression
+        filename = getattr(uri, 'name', 'unknown')
         return uri, filename
     else:
-        raise TypeError('don\'t know how to handle uri %s' % repr(uri))
+        raise TypeError("don't know how to handle uri %r in mode %r" % (uri, mode))
```

Comment:
I don't think the mode is relevant here. This if block doesn't check the mode, it only checks the uri.

Reply:
You need to agree with @piskvorky here https://github.com/RaRe-Technologies/smart_open/pull/263#discussion_r259590067

Reply:
@piskvorky, thoughts?

Reply:
The previous version used complex URI checking, which depended on the `mode`, hence the change in logging.

If the current version doesn't rely on `mode` any more (reverted), I guess we don't need to show it (but probably no harm either).

Reply:
My motivation for removing the mode is that keeping it misleads the user into thinking mode is part of the problem. In this particular case, it isn’t.